x/programs/examples/imports/pstate/pstate.go

```diff
@@ -0,0 +1,182 @@
+}
+
+func (i *Import) Name() string {
+	return Name
+}
+
+func (i *Import) Register(link runtime.Link, meter runtime.Meter, _ runtime.SupportedImports) error {
+	if i.registered {
+		return fmt.Errorf("import module already registered: %q", Name)
+	}
+	i.meter = meter
+	i.registered = true
+
+	if err := link.FuncWrap(Name, "put", i.putFn); err != nil {
+		return err
+	}
+	if err := link.FuncWrap(Name, "get", i.getFn); err != nil {
+		return err
+	}
+	if err := link.FuncWrap(Name, "len", i.getLenFn); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (i *Import) putFn(caller *wasmtime.Caller, idPtr int64, keyPtr int32, keyLength int32, valuePtr int32, valueLength int32) int32 {
+	memory := runtime.NewMemory(runtime.NewExportClient(caller))
+	programIDBytes, err := memory.Range(uint64(idPtr), uint64(ids.IDLen))
+	if err != nil {
+		i.log.Error("failed to read program id from memory",
+			zap.Error(err),
+		)
+		return -1
+	}
+
+	keyBytes, err := memory.Range(uint64(keyPtr), uint64(keyLength))
+	if err != nil {
+		i.log.Error("failed to read key from memory",
+			zap.Error(err),
+		)
+		return -1
+	}
+
+	valueBytes, err := memory.Range(uint64(valuePtr), uint64(valueLength))
+	if err != nil {
+		i.log.Error("failed to read value from memory",
+			zap.Error(err),
+		)
+		return -1
+	}
+
+	k := storage.ProgramPrefixKey(programIDBytes, keyBytes)
+	err = i.mu.Insert(context.Background(), k, valueBytes)
+	if err != nil {
+		i.log.Error("failed to insert into storage",
+			zap.Error(err),
+		)
+		return -1
+	}
+
+	return 0
+}
+
+func (i *Import) getLenFn(caller *wasmtime.Caller, idPtr int64, keyPtr int32, keyLength int32) int32 {
+	memory := runtime.NewMemory(runtime.NewExportClient(caller))
+	programIDBytes, err := memory.Range(uint64(idPtr), uint64(ids.IDLen))
+	if err != nil {
+		i.log.Error("failed to read program id from memory",
+			zap.Error(err),
+		)
+		return -1
+	}
+
+	keyBytes, err := memory.Range(uint64(keyPtr), uint64(keyLength))
+	if err != nil {
+		i.log.Error("failed to read key from memory",
+			zap.Error(err),
+		)
+		return -1
+	}
+
+	k := storage.ProgramPrefixKey(programIDBytes, keyBytes)
+	val, err := i.mu.GetValue(context.Background(), k)
+	if err != nil {
+		if !errors.Is(err, database.ErrNotFound) {
+			i.log.Error("failed to get value from storage",
+				zap.Error(err),
+			)
+		}
+		return -1
+	}
+
+	return int32(len(val))
+}
+
+func (i *Import) getFn(caller *wasmtime.Caller, idPtr int64, keyPtr int32, keyLength int32, valLength int32) int32 {
+	memory := runtime.NewMemory(runtime.NewExportClient(caller))
+	programIDBytes, err := memory.Range(uint64(idPtr), uint64(ids.IDLen))
+	if err != nil {
+		i.log.Error("failed to read program id from memory",
+			zap.Error(err),
+		)
+		return -1
+	}
+
+	keyBytes, err := memory.Range(uint64(keyPtr), uint64(keyLength))
+	if err != nil {
+		i.log.Error("failed to read key from memory",
+			zap.Error(err),
+		)
+		return -1
+	}
+
+	k := storage.ProgramPrefixKey(programIDBytes, keyBytes)
```

Comment:
This will also need to have the max size (in chunks) suffixed at the end of it but we can do that in a follow-up.

Reply:
https://github.com/ava-labs/hypersdk/blob/main/keys/keys.go

Reply:
https://github.com/ava-labs/hypersdk/issues/528